### DISCUSSION TITLE
Generating an SBOM for a PyInstaller package

### DISCUSSION THREAD

**Discussion by matth65** [pyinstaller/pyinstaller#7821] [PyInstaller]
Is there way to generate an sbom (software bill of materials) for a PyInstaller package that would include the pip dependencies from requirements.txt and the list of dll/so files pulled in during the build process? It is hard to tell from the TOC files what python/pip packages are actually included in the packaged binary output.

Using pip freeze from the venv to generate the python dependencies is a little difficult because the PyInstaller build dependencies are mixed in with my application dependencies.  I believe the following pip3 packages are included as build dependencies for PyInstaller (not part of my application): altgraph, pyinstaller, pyinstaller-hooks-contrib, which are not required for running my application.

```
$ pip3 freeze
altgraph==0.17.3
dnspython==2.4.1
fastjsonschema==2.16.2
psutil==5.9.4
pyinstaller==5.7.0
pyinstaller-hooks-contrib==2023.6
pyroute2==0.7.3
```

My sbom would include dnspython, fastjsonschema, psutil, pyroute2 from the example above.

Best Regards,
Matt

**Comment by rokm** [comment]
Out of the box, no.

But I imagine you could come up with a script that takes the TOC and cross-reference the source paths of collected files with contents of the packages installed in the environment.

**Reply by matth65** [reply]
The Analysis.toc file seems to include more than that is contained in the binary package output. Which of the TOC files would provide a list of the python/pip dependencies that are contained in the packaged binary output? Can I use the same TOC file to determine which dll/so files are included?

**Reply by rokm** [reply]
Analysis.toc is actually a literal representation of a tuple in which "guts" of Analysis object are collected (i.e., you can read the file and `eval` it; or use `PyInstaller.utils.misc.load_py_data_struct`, which does the same). E.g., for v5.13:
https://github.com/pyinstaller/pyinstaller/blob/58b8235b0ecd474f71c56e119d2889bf09026cb4/PyInstaller/building/build_main.py#L431-L461

So you should be interested in lists that correspond to `pure` (pure python module), `binaries` (DLLs) and `datas` (other data files). So in elements at indices 15, 16, and 19 in the tuple from the Analysis-00.toc file. 

The entries from those lists are all collected into the frozen application; however, while binaries and data files are collected as files (in onedir mode), pure python modules are always collected into embedded PYZ archive (so you don't see them).
